packages/api-page-builder/src/graphql/graphql/pages.gql.ts

```diff
@@ -285,17 +285,41 @@ const createBasePageGraphQL = (): GraphQLSchemaPlugin<PbContext> => {
                         }
 
                         const blocks = [];
-                        for (const block of page.content?.elements) {
-                            const blockId = block.data?.blockId;
+                        for (const pageBlock of page.content?.elements) {
+                            const blockId = pageBlock.data?.blockId;
+                            // If block has blockId, then it is reference block and we need to get elements for it
                             if (blockId) {
                                 const blockData = await context.pageBuilder.getPageBlock(blockId);
+                                // We check if block has variable values set on the page and use them in priority over ones,
+                                // that are set in blockEditor
+                                const blockDataVariables =
+                                    blockData?.content?.data?.variables || [];
+                                const variables = blockDataVariables.map(
+                                    (blockDataVariable: any) => {
+                                        const value =
+                                            pageBlock.data?.variables?.find(
+                                                (variable: any) =>
+                                                    variable.varRef === blockDataVariable.varRef
+                                            )?.value || blockDataVariable.value;
+
+                                        return {
+                                            ...blockDataVariable,
+                                            value
+                                        };
+                                    }
+                                );
+
                                 blocks.push({
-                                    ...block,
-                                    data: { blockId, ...blockData?.content?.data },
+                                    ...pageBlock,
+                                    data: {
+                                        blockId,
+                                        ...blockData?.content?.data,
+                                        variables
+                                    },
                                     elements: blockData?.content?.elements || []
                                 });
                             } else {
-                                blocks.push(block);
+                                blocks.push(pageBlock);
```

Comment:
Please extract this logic into a separate file, just so we keep the resolvers shorter and easier to read. Besides, you'll be able to unit test this mapping logic if it lives in a standalone function.